lib/AST/Decl.cpp

```diff
@@ -1296,20 +1296,34 @@ static bool isPolymorphic(const AbstractStorageDecl *storage) {
 /// Determines the access semantics to use in a DeclRefExpr or
 /// MemberRefExpr use of this value in the specified context.
 AccessSemantics
-ValueDecl::getAccessSemanticsFromContext(const DeclContext *UseDC) const {
+ValueDecl::getAccessSemanticsFromContext(const DeclContext *UseDC,
+                                         Expr *base) const {
   // If we're inside a @_transparent function, use the most conservative
   // access pattern, since we may be inlined from a different resilience
   // domain.
   ResilienceExpansion expansion = UseDC->getResilienceExpansion();
 
   if (auto *var = dyn_cast<AbstractStorageDecl>(this)) {
-    // Observing member are accessed directly from within their didSet/willSet
-    // specifiers.  This prevents assignments from becoming infinite loops.
-    if (auto *UseFD = dyn_cast<AccessorDecl>(UseDC))
+    // Prevent variable mutations from within their own didSet/willSet
+    // specifiers from becoming infinite loops by accessing directly.
+    if (auto *UseFD = dyn_cast<AccessorDecl>(UseDC)) {
       if (var->hasStorage() && var->hasAccessorFunctions() &&
-          UseFD->getStorage() == var)
-        return AccessSemantics::DirectToStorage;
-    
+          UseFD->getStorage() == var) {
+        // A plain variable access from within its own observer is direct.
+        if (!base)
+          return AccessSemantics::DirectToStorage;
+
+        // A member access on the implicit 'self' declaration within its own
+        // observer is direct. If the base is some other expression, we want
+        // to call the setter, as we might be accessing the member on a
+        // *different* instance.
+        base = base->getValueProvidingExpr();
+        if (auto baseDRE = dyn_cast<DeclRefExpr>(base))
+          if (baseDRE->getDecl() == UseFD->getImplicitSelfDecl())
```

Comment:
This probably won't work if the base expression references a 'self' decl in a capture list and not the actual 'self'.

Reply:
@slavapestov Ah, good point! Looking into it, we actually already don't perform direct accesses to members accessed in closures within their own observers (as we only look at `AccessorDecl` decl contexts), e.g:
```swift
struct S {
  var i = 0 {
    didSet {
      { self.i = 1 }()
    }
  }
}

var s = S()
s.i = 9 // already infinite recursion
```

This PR won't change that behaviour. Would it be worth looking into allowing direct access in this case in this PR, or should this be done in a separate PR?

Reply:
I think the current behavior is correct, just like in initializers we go through the property instead of doing direct access when within a closure.